src/ConfigWidgets/SymbolsDialog.cpp

```diff
@@ -4,36 +4,54 @@
 
 #include "ConfigWidgets/SymbolsDialog.h"
 
+#include <absl/strings/str_format.h>
+
 #include <QFileDialog>
 #include <QMessageBox>
 #include <QPushButton>
 #include <QSettings>
 #include <algorithm>
+#include <filesystem>
+#include <iterator>
 #include <memory>
+#include <tuple>
 
+#include "GrpcProtos/module.pb.h"
+#include "ObjectUtils/SymbolsFile.h"
+#include "OrbitBase/Logging.h"
+#include "OrbitBase/Result.h"
 #include "ui_SymbolsDialog.h"
 
 constexpr const char* kFileDialogSavedDirectoryKey = "symbols_file_dialog_saved_directory";
+constexpr const char* kModuleHeadlineLabel = "Add Symbols for <font color=\"red\">%1</font>";
+
+using orbit_client_data::ModuleData;
+using orbit_grpc_protos::ModuleInfo;
 
 namespace orbit_config_widgets {
 
-SymbolsDialog::~SymbolsDialog() = default;
+SymbolsDialog::~SymbolsDialog() { persistent_storage_manager_->SavePaths(GetSymbolPaths()); }
 
-SymbolsDialog::SymbolsDialog(QWidget* parent)
-    : QDialog(parent), ui_(std::make_unique<Ui::SymbolsDialog>()) {
+SymbolsDialog::SymbolsDialog(
+    orbit_symbol_paths::PersistentStorageManager* persistent_storage_manager,
+    std::optional<const ModuleData*> module, QWidget* parent)
+    : QDialog(parent),
+      ui_(std::make_unique<Ui::SymbolsDialog>()),
+      module_(module),
+      persistent_storage_manager_(persistent_storage_manager) {
+  ORBIT_CHECK(persistent_storage_manager_ != nullptr);
   ui_->setupUi(this);
 
-  QObject::connect(ui_->addButton, &QPushButton::clicked, this, &SymbolsDialog::OnAddButtonClicked);
+  // Initially disabled, will be enabled once selection changes.
+  ui_->removeButton->setEnabled(false);
 
-  QObject::connect(ui_->removeButton, &QPushButton::clicked, this,
-                   &SymbolsDialog::OnRemoveButtonClicked);
+  SetSymbolPaths(persistent_storage_manager_->LoadPaths());
```

Comment:
I looked at it more from this direction: The SymbolsDialog has now the done button instead of cancel/add and conceptually is responsible for loading and saving. The user/caller of SymbolsDialog does not need to remember load the data into the dialog and take it out again. And there is more functionality planned for saving the direct mappings from module to symbols, which is not just a simple path. Also, I dont think the complexity of the tests should dictate the design of the component.. The mock code is a bit overhead, but also not to bad. wdyt?

Reply:
I think I can live with it.